### PR TITLE
Python3 port for test_networkgroup.py, test_protocol.py

### PR DESCRIPTION
**Pull request by kdcis**
Fix Imports
Contains changes from #1786 for protocol.py as it needed to run test_networkgroup
Handled bytes to string comparison/concatenation cases
Python 2/3 compatibility

### CODE REVIEW

File: src/highlevelcrypto.py [hunk 1]
```diff
@@ -9,10 +9,16 @@
 
 from binascii import hexlify
 
-import pyelliptic
-from bmconfigparser import BMConfigParser
-from pyelliptic import OpenSSL
-from pyelliptic import arithmetic as a
+try:
```
Comment (PeterSurda) [inline src/highlevelcrypto.py:12]: I'm not sure about this. Need to think about it.

File: src/protocol.py [hunk 2]
```diff
@@ -13,16 +13,30 @@
...
-import highlevelcrypto
-import state
-from addresses import (
-    encodeVarint, decodeVarint, decodeAddress, varintDecodeError)
-from bmconfigparser import BMConfigParser
-from debug import logger
-from fallback import RIPEMD160Hash
-from helper_sql import sqlExecute
-from version import softwareVersion
+try:
```
Comment (PeterSurda) [inline src/protocol.py:16]: I'm also not sure about these imports.

File: src/protocol.py [hunk 3]
```diff
@@ -13,16 +13,30 @@
...
+    from pybitmessage import state
+    from pybitmessage.addresses import (
+        encodeVarint, decodeVarint, decodeAddress, varintDecodeError)
+    from pybitmessage.bmconfigparser import BMConfigParser
+    from pybitmessage.debug import logger
+    from pybitmessage.fallback import RIPEMD160Hash
+    from pybitmessage.helper_sql import sqlExecute
+    from pybitmessage.version import softwareVersion
+
+PY3 = sys.version_info[0] >= 3
```
Comment (PeterSurda) [inline src/protocol.py:39]: This isn't needed, use `six.PY2`.

File: src/protocol.py [hunk 4]
```diff
@@ -90,19 +104,30 @@ def isBitSetWithinBitfield(fourByteString, n):
...
-def encodeHost(host):
-    """Encode a given host to be used in low-level socket operations"""
-    if host.find('.onion') > -1:
-        return '\xfd\x87\xd8\x7e\xeb\x43' + base64.b32decode(
-            host.split(".")[0], True)
-    elif host.find(':') == -1:
-        return '\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xFF\xFF' + \
-            socket.inet_aton(host)
-    return socket.inet_pton(socket.AF_INET6, host)
+if PY3:
```
Comment (PeterSurda) [inline src/protocol.py:110]: Here I would refactor the original code. Create new constants for the raw strings ,and then you can keep the old codepath inside the function. Like this:

```
if sx.PY2:
    ONION_PREFIX = '\xfd\x87\xd8\x7e\xeb\x43'
else:
    ONION_PREFIX = b'\xfd\x87\xd8\x7e\xeb\x43'
```

File: src/protocol.py [hunk 5]
```diff
@@ -142,59 +167,116 @@ def network_group(host):
...
-            logger.debug(
-                'Ignoring IP address in private range: %s', hostStandardFormat)
-        return hostStandardFormat if private else False
-    if host[0:2] >= '\xAC\x10' and host[0:2] < '\xAC\x20':  # 172.16/12
-        if not private:
-            logger.debug(
-                'Ignoring IP address in private range: %s', hostStandardFormat)
-        return hostStandardFormat if private else False
-    return False if private else hostStandardFormat
+if PY3:
```
Comment (PeterSurda) [inline src/protocol.py:170]: Same as with `encodeHost`: refactor to use a constant and `six`. The constants are the same as with `encodeHost`, no need to duplicate.

File: src/protocol.py [hunk 6]
```diff
@@ -142,59 +167,116 @@ def network_group(host):
...
+            except ValueError:
+                return False
+            if hostStandardFormat == "":
+                # This can happen on Windows systems which are
+                # not 64-bit compatible so let us drop the IPv6 address.
+                return False
+            return checkIPv6Address(host, hostStandardFormat, private)
+
+
+if PY3:
```
Comment (PeterSurda) [inline src/protocol.py:223]: Same as above, refactor using constants and `six`.

File: src/protocol.py [hunk 7]
```diff
@@ -210,7 +292,7 @@ def checkIPv6Address(host, hostStandardFormat, private=False):
         if not private:
             logger.debug('Ignoring local address: %s', hostStandardFormat)
         return hostStandardFormat if private else False
-    if (ord(host[0]) & 0xfe) == 0xfc:
+    if (ord(str(host[0])) & 0xfe) == 0xfc:
```
Comment (PeterSurda) [inline src/protocol.py:295]: I don't know if this is ok as there are no tests for this. So you need to write a test.

File: src/tests/common.py [hunk 8]
```diff
@@ -13,7 +13,10 @@
 def cleanup(home=None, files=_files):
     """Cleanup application files"""
     if not home:
-        import state
+        try:
```
Comment (PeterSurda) [inline src/tests/common.py:16]: Again I'm not sure about this.

File: src/tests/test_networkgroup.py [hunk 9]
```diff
@@ -2,30 +2,45 @@
 Test for network group
 """
 import unittest
+import sys
```
Comment (PeterSurda) [inline src/tests/test_networkgroup.py:5]: You don't need `sys`, use `six` instead.

File: src/tests/test_networkgroup.py [hunk 10]
```diff
@@ -2,30 +2,45 @@
...
+try:
```
Comment (PeterSurda) [inline src/tests/test_networkgroup.py:6]: Again not sure, I need to think about it.

File: src/tests/test_networkgroup.py [hunk 11]
```diff
@@ -2,30 +2,45 @@
...
+    def tearDown(self):
+        cleanup()
+
     def test_network_group(self):
         """Test various types of network groups"""
         from pybitmessage.protocol import network_group
 
         test_ip = '1.2.3.4'
-        self.assertEqual('\x01\x02', network_group(test_ip))
+        if PY3:
```
Comment (PeterSurda) [inline src/tests/test_networkgroup.py:27]: Use constants or class variables.

File: src/tests/test_networkgroup.py [hunk 12]
```diff
@@ -2,30 +2,45 @@
...
+            self.assertEqual('\x01\x02', network_group(test_ip))
 
         test_ip = '127.0.0.1'
         self.assertEqual('IPv4', network_group(test_ip))
 
         test_ip = '0102:0304:0506:0708:090A:0B0C:0D0E:0F10'
-        self.assertEqual(
-            '\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0A\x0B\x0C',
-            network_group(test_ip))
+        if PY3:
```
Comment (PeterSurda) [inline src/tests/test_networkgroup.py:36]: same here.

**Review by PeterSurda — CHANGES_REQUESTED** [review]

File: src/protocol.py (hunk 7 above)
Comment (kdcis) [reply src/protocol.py:248]: Okay, I'll write one

File: src/highlevelcrypto.py (hunk 1 above)
Comment (g1itch) [reply src/highlevelcrypto.py:12]: You don't need this, please look into #1788

Comment (g1itch) [reply src/highlevelcrypto.py:12]: However in tests the imports gonna be like these. See: a8c79c3832e2e3081ebdcb0373f6b4c0057906f7.

Comment (kdcis) [reply src/highlevelcrypto.py:12]: Okay, got it

File: src/protocol.py (hunk 7 above)
Comment (g1itch) [reply src/protocol.py:252]: I've already solved this in 7e1f6e8. Yes, at first I wrote more tests.

File: src/protocol.py (hunk 4 above)
Comment (g1itch) [reply src/protocol.py]: python2.7 has no problem with bytes literals

File: src/protocol.py (hunk 7 above)
Comment (kdcis) [reply src/protocol.py:240]: Okay, I'll merge your changes

**Comment by kdcis** [comment]
Closing. Fixed via #1792

**Review by PeterSurda — CHANGES_REQUESTED** [review]

File: src/protocol.py [hunk 13]
```diff
@@ -161,7 +161,7 @@ def checkIPAddress(host, private=False):
             hostStandardFormat = socket.inet_ntop(socket.AF_INET6, host)
         except ValueError:
             return False
-        if hostStandardFormat == "":
+        if len(hostStandardFormat) == 0:
```
[the remark this answers is not in the document]
Comment (kdcis) [reply src/protocol.py:164]: It works

[the remark this answers is not in the document]
Comment (g1itch) [reply src/protocol.py:164]: it's about str vs bytes here, to not insert a literal like `b''`, maybe bool makes sence but the condition is rather about the emptiness of the `hostStandardFormat` value as I see it

File: src/protocol.py [hunk 14]
```diff
@@ -327,13 +331,13 @@ def assembleVersionMessage(
...
-        (NODE_SSL if haveSSL(server) else 0) |
-        (NODE_DANDELION if state.dandelion else 0)
+        NODE_NETWORK
+        | (NODE_SSL if haveSSL(server) else 0)
+        | (NODE_DANDELION if state.dandelion else 0)
     )
     # = 127.0.0.1. This will be ignored by the remote host.
     # The actual remote connected IP will be used.
-    payload += '\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xFF\xFF' + pack(
+    payload += b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xFF\xFF' + pack(
```
[the remark this answers is not in the document]
Comment (kdcis) [reply src/protocol.py:340]: No it does not on python2,

It returns '\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff\x7f\x00\x00\x01'

File: src/tests/test_networkgroup.py [hunk 15]
```diff
@@ -2,42 +2,35 @@
...
+from .common import cleanup
 
 
 class TestNetworkGroup(unittest.TestCase):
     """
     Test case for network group
     """
+
+    def tearDown(self):
+        cleanup()
```
Comment (g1itch) [inline src/tests/test_networkgroup.py:14]: I don't think you need this here. Can you show a corresponding failing check?

Comment (kdcis) [reply src/tests/test_networkgroup.py:14]: I was getting this error ```'Found application file keys.dat in src dir'```

So, I added tearDown() that fixed it

Comment (kdcis) [reply src/tests/test_networkgroup.py:14]: I will retest removing it, once more

File: src/tests/common.py [hunk 16]
```diff
@@ -13,7 +13,7 @@
 def cleanup(home=None, files=_files):
     """Cleanup application files"""
     if not home:
-        import state
+        from pybitmessage import state
```
Comment (g1itch) [inline src/tests/common.py]: you've already got the pathmagic for this: the `common` module is used in both test suites
 - in the external tests we write `from pybitmessage import ..`
 - but it the core tests it's still `import state`

I suppose we gonna switch to `from . import state` inside of the `pybitmessage` package and using another start script in top level dir, but not earlier than the circular import in the `network` package is solved.

File: src/protocol.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (g1itch) [reply src/protocol.py:340]: what's called bytes in py3 in python2 is str
the py3 str analogue in python2 is unicode

[the remark this answers is not in the document]
Comment (g1itch) [reply src/protocol.py:340]: that's why you are not gonna get a portable code if you write `str(data)`, instead you do `data.encode('utf-8')  # arg may be dropped in python3`

File: src/tests/test_networkgroup.py (hunk 15 above)
Comment (g1itch) [reply src/tests/test_networkgroup.py:14]: please post the link

Comment (kdcis) [reply src/tests/test_networkgroup.py:14]: https://github.com/Bitmessage/PyBitmessage/pull/1792/checks?sha=8d545fa066cdda420e1450ca248dc79ee1f74fba

I can't find the link for that logs, but it was for this commit

File: src/tests/test_networkgroup.py [hunk 17]
```diff
@@ -2,42 +2,46 @@
...
-        self.assertEqual(
-            test_ip,
-            network_group(test_ip))
+        self.assertEqual(test_ip, network_group(test_ip))
 
         test_ip = None
-        self.assertEqual(
-            None,
-            network_group(test_ip))
+        self.assertEqual(None, network_group(test_ip))
```
[the remark this answers is not in the document]
Comment (g1itch) [reply src/tests/test_networkgroup.py:36]: I can't understand, why you have kept this? There is another copy of this test with additional samples in the `test_protocol` module
```
                                     |
                                     |
                                     V
```

File: src/tests/test_networkgroup.py [hunk 18]
```diff
@@ -2,42 +2,46 @@
 Test for network group
 """
 import unittest
-
-from .common import skip_python3
-
-skip_python3()
+import six
+try:
+    from .common import cleanup
```
[the remark this answers is not in the document]
Comment (kdcis) [reply src/tests/test_networkgroup.py]: Worked

File: src/tests/test_networkgroup.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (kdcis) [reply src/tests/test_networkgroup.py:36]: Worked

File: src/tests/test_protocol.py [hunk 19]
```diff
@@ -2,25 +2,97 @@
...
+    @unittest.skipIf(
+        sys.hexversion >= 0x3000000, 'this is still not working with python3')
+    def test_check_local_socks(self):
+        """The SOCKS part of the local check"""
         self.assertTrue(
             not protocol.checkSocksIP('127.0.0.1')
             or state.socksIP)
+
+    def test_network_group(self):
+        """Test various types of network groups"""
```
Comment (g1itch) [inline src/tests/test_protocol.py:65]: This is the new `test_network_group()`

File: src/tests/test_networkgroup.py (hunk 15 above)
Comment (g1itch) [reply src/tests/test_networkgroup.py:14]: Probably it's because you should've remove this module, not add another bikeshed, see. below.

**Review by g1itch — CHANGES_REQUESTED** [review]
You've lost the logic of the tests refactoring. network group test goes to `test_protocol`

File: src/tests/test_blindsig.py [hunk 20]
```diff
@@ -1,17 +1,16 @@
+# pylint: disable=protected-access
 """
 Test for ECC blind signatures
 """
+from pybitmessage.pyelliptic import ECCBlind, ECCBlindChain, OpenSSL
+
```
Comment (g1itch) [inline src/tests/test_blindsig.py:6]: wrong order

File: src/tests/common.py [hunk 21]
```diff
@@ -13,7 +14,6 @@
 def cleanup(home=None, files=_files):
     """Cleanup application files"""
     if not home:
-        import state
```
Comment (g1itch) [inline src/tests/common.py:16]: You don't need this hunk either

File: src/tests/test_blindsig.py (hunk 20 above)
Comment (kdcis) [reply src/tests/test_blindsig.py:6]: There was code quality error popped up saying module level imports should be on top

Comment (g1itch) [reply src/tests/test_blindsig.py:6]: nonsense: this is a library import, why did you place it before the standard library ones and why the for- import is before simple ones? you recently posted a link to pep8 ... so go read it yourself

File: src/tests/test_protocol.py (hunk 19 above)
Comment (kdcis) [reply src/tests/test_protocol.py:65]: Okay, so I guess there's no need for test_networkgroup.py? 
Should we remove it?

File: src/tests/test_blindsig.py (hunk 20 above)
Comment (kdcis) [reply src/tests/test_blindsig.py]: Ok, I reverted the changes
https://github.com/Bitmessage/PyBitmessage/pull/1792/checks?check_run_id=3230496739

Comment (g1itch) [reply src/tests/test_blindsig.py]: this branch is out of date, rebase

File: src/tests/test_api.py [hunk 22]
```diff
@@ -7,15 +7,15 @@
...
     from xmlrpclib import ServerProxy, ProtocolError
 except ImportError:
     from xmlrpc.client import ServerProxy, ProtocolError
 
 from .test_process import TestProcessProto, TestProcessShutdown
 
+skip_python3()
+
 
 class TestAPIProto(TestProcessProto):
```
Comment (g1itch) [inline src/tests/test_api.py:20]: This hunk was changed in v0.6

File: src/tests/test_blindsig.py [hunk 23]
```diff
@@ -6,12 +6,12 @@
...
 from .common import skip_python3
 
-skip_python3()
-
 from pybitmessage.pyelliptic import ECCBlind, ECCBlindChain, OpenSSL
 
 # pylint: disable=protected-access
 
+skip_python3()
+
```
Comment (g1itch) [inline src/tests/test_blindsig.py:14]: there is no `skip_python3()` in v0.6, this hunk is not fit

File: src/tests/test_blindsig.py (hunk 20 above)
Comment (kdcis) [reply src/tests/test_blindsig.py]: Ok I found the issue, It was not properly rebased. I fixed it now.

File: src/tests/test_blindsig.py (hunk 23 above)
Comment (kdcis) [reply src/tests/test_blindsig.py]: Yep, Fixed now

File: src/tests/common.py [hunk 24]
```diff
@@ -13,7 +13,10 @@
...
+        try:
+            import state
```
[the remark this answers is not in the document]
Comment (kdcis) [reply src/tests/common.py]: reverted

File: src/protocol.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (g1itch) [reply src/protocol.py:340]: $ python
Python 2.7.18 (default, Aug 26 2020, 16:28:25) 
[GCC 9.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> b'x7f\x00\x00\x01' == 'x7f\x00\x00\x01'
True
>>>

**Review by g1itch — APPROVED** [review]
OK. Now you need a Peter's review. I see this as a good material for squashing (;

**Review by PeterSurda — APPROVED** [review]

**Comment by g1itch** [comment]
Bad signature as well here

**Comment by kdcis** [comment]
> Bad signature as well here

Updated verified commit.

**Comment by g1itch** [comment]
> > Bad signature as well here
> 
> Updated verified commit.

Hmm, it looks like you've changed a GPG key...

**Comment by kdcis** [comment]
> > > Bad signature as well here
> > 
> > 
> > Updated verified commit.
> 
> Hmm, it looks like you've changed a GPG key...

Yes.

**Review by g1itch — CHANGES_REQUESTED** [review]
Please rebase and change the commit message.

**Pull request closed without merging**